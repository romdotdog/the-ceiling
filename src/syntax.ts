import { assert } from "./util.js";

// prettier-ignore
class StringScanner {
  constructor(private src: string, public p: number) {}

  private char(): string | undefined {
    return this.src[this.p];
  }

  private readHexDigits(n: number): string {
    let result = "";
    for (let i = 0; i < n; i++) {
      const ch = this.char();
      if (ch === undefined || !/[0-9a-fA-F]/.test(ch)) {
        throw new Error(`invalid hex escape at ${this.p}`);
      }
      result += ch;
      this.p += 1;
    }
    return result;
  }

  private readUnicodeEscape(): string {
    this.p += 1; // skip 'u'
    if (this.char() === "{") {
      this.p += 1; // skip {
      let hex = "";
      while (true) {
        const ch = this.char();
        if (ch === undefined) throw new Error("unterminated unicode escape");
        if (ch === "}") { this.p += 1; break; }
        if (!/[0-9a-fA-F]/.test(ch)) throw new Error(`invalid unicode at ${this.p}`);
        hex += ch;
        this.p += 1;
      }
      return String.fromCodePoint(parseInt(hex, 16));
    }
    return String.fromCharCode(parseInt(this.readHexDigits(4), 16));
  }

  private readHexEscape(): string {
    this.p += 1; // skip 'x'
    return String.fromCharCode(parseInt(this.readHexDigits(2), 16));
  }

  public scanString(): string {
    const quote = this.char();
    if (quote !== '"' && quote !== "'") {
      throw new Error(`expected string literal at ${this.p}`);
    }
    this.p += 1; // skip opening quote

    let value = "";
    while (true) {
      const ch = this.char();
      if (ch === undefined) throw new Error("unterminated string literal");
      if (ch === quote) { this.p += 1; break; }
      if (ch === "\\") {
        this.p += 1;
        const esc = this.char();
        if (esc === undefined) throw new Error("bad escape at <eof>");
        switch (esc) {
          case "n": value += "\n"; this.p += 1; break;
          case "r": value += "\r"; this.p += 1; break;
          case "t": value += "\t"; this.p += 1; break;
          case "\\": value += "\\"; this.p += 1; break;
          case "\"": value += "\""; this.p += 1; break;
          case "'": value += "'"; this.p += 1; break;
          case "u": value += this.readUnicodeEscape(); break;
          case "x": value += this.readHexEscape(); break;
          default: value += esc; this.p += 1; break;
        }
      } else {
        value += ch;
        this.p += 1;
      }
    }

    return value;
  }
}

export enum TokenKind {
  Equals,
  Colon,
  Semicolon,
  Number,
  String,
  Identifier,
  Const,
  Let,
}

type Token =
  | { kind: Exclude<TokenKind, TokenKind.Number | TokenKind.String>; start: number; end: number }
  | { kind: TokenKind.Number; start: number; end: number; value: number }
  | { kind: TokenKind.String; start: number; end: number; value: string };

export class Lexer {
  private p = 0;
  constructor(private src: string) {}

  private done() {
    return this.p >= this.src.length;
  }

  private char(): string | undefined {
    return this.src[this.p];
  }

  private skipWhitespace() {
    while (true) {
      switch (this.char()) {
        case " ":
        case "\t":
        case "\r":
        case "\n":
          this.p += 1;
          break;
        default:
          return;
      }
    }
  }

  private token(): Token | null {
    const start = this.p;
    switch (this.char()) {
      case "=":
        this.p += 1;
        return { kind: TokenKind.Equals, start, end: this.p };
      case ":":
        this.p += 1;
        return { kind: TokenKind.Colon, start, end: this.p };
      case ";":
        this.p += 1;
        return { kind: TokenKind.Semicolon, start, end: this.p };
      case "'":
      case '"': {
        const scanner = new StringScanner(this.src, this.p);
        const value = scanner.scanString();
        const end = scanner.p;
        this.p = end;
        return { kind: TokenKind.String, start, end, value };
      }
    }
    return null;
  }

  public *run(): Generator<Token, void, unknown> {
    while (true) {
      this.skipWhitespace();
      const maybeToken = this.token();
      if (maybeToken) {
        yield maybeToken;
        continue;
      }
      if (this.done()) break;

      // start taking an identifier
      const start = this.p;
      ident: while (true) {
        switch (this.char()) {
          case undefined:
          case " ":
          case "\t":
          case "\r":
          case "\n":
            yield this.processIdent(start, this.p);
            break ident;
          default: {
            const maybeEnd = this.p;
            const maybeToken = this.token();
            if (maybeToken) {
              yield this.processIdent(start, maybeEnd);
              yield maybeToken;
              break ident;
            }
            this.p += 1;
          }
        }
      }
    }
  }

  private processIdent(start: number, end: number): Token {
    const content = this.src.slice(start, end);
    switch (content) {
      case "let":
        return { kind: TokenKind.Let, start, end };
      case "const":
        return { kind: TokenKind.Const, start, end };
      case "NaN":
        return { kind: TokenKind.Number, start, end, value: NaN };
    }
    const float = Number(content);
    if (!isNaN(float)) {
      return { kind: TokenKind.Number, start, end, value: float };
    }
    return { kind: TokenKind.Identifier, start, end };
  }
}

class Parser {
  private lexer: Generator<Token, void, unknown>;
  private token: Token | null = null;
  constructor(lexer: Lexer) {
    this.lexer = lexer.run();
    this.next();
  }

  private next() {
    const { value, done } = this.lexer.next();
    if (done) {
      this.token = null;
    } else {
      assert(value !== undefined, "value should not be undefined");
      this.token = value;
    }
  }
}
