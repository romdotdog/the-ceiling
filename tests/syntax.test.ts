import { expect, it } from "vitest";
import { Lexer, TokenKind } from "../src/syntax.js";

interface ReadableToken {
  kind: string;
  source: string;
  value?: any;
}

function lex(src: string): ReadableToken[] {
  const tokens = [];
  const lexer = new Lexer(src);
  for (const token of lexer.run()) {
    tokens.push({
      kind: TokenKind[token.kind],
      source: src.slice(token.start, token.end),
      ...("value" in token ? { value: token.value } : {}),
    });
  }
  return tokens;
}

it("let hello = 1;", () => {
  expect(lex("let hello = 1;")).toMatchSnapshot();
});

it("const foo = 42;", () => {
  expect(lex("const foo = 42;")).toMatchSnapshot();
});

it("identifiers and colons", () => {
  expect(lex("foo:bar;")).toMatchSnapshot();
});

it("floating numbers", () => {
  expect(lex("let pi = 3.14;")).toMatchSnapshot();
});

it("NaN literal", () => {
  expect(lex("let x = NaN;")).toMatchSnapshot();
});

it("string single quotes", () => {
  expect(lex("let s = 'hello';")).toMatchSnapshot();
});

it("string double quotes", () => {
  expect(lex('let s = "world";')).toMatchSnapshot();
});

it("string escapes", () => {
  expect(lex("'a\\nb'")).toMatchSnapshot();
});

it("whitespace and newlines", () => {
  expect(lex("\n\t  let   x=1\n\n")).toMatchSnapshot();
});

it("unquoted identifier with digits", () => {
  expect(lex("foo123")).toMatchSnapshot();
});

it("multiple identifiers", () => {
  expect(lex("a b c")).toMatchSnapshot();
});

it("mixed operators", () => {
  expect(lex("a=b:c;")).toMatchSnapshot();
});

it("unicode identifiers", () => {
  expect(lex("let α = 1; const π = 3.14;")).toMatchSnapshot();
  expect(lex("let язык = 'lang';")).toMatchSnapshot();
  expect(lex("let \u0301 = 'é';")).toMatchSnapshot();
  expect(lex("let ❌ = 1;")).toMatchSnapshot();
});
