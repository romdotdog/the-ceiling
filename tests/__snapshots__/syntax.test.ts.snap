// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NaN literal 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "x",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "NaN",
    "value": NaN,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`const foo = 42; 1`] = `
[
  {
    "kind": "Const",
    "source": "const",
  },
  {
    "kind": "Identifier",
    "source": "foo",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "42",
    "value": 42,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`floating numbers 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "pi",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "3.14",
    "value": 3.14,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`greek identifiers 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "α",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "1",
    "value": 1,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
  {
    "kind": "Const",
    "source": "const",
  },
  {
    "kind": "Identifier",
    "source": "π",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "3.14",
    "value": 3.14,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`greek identifiers 2`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "язык",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "String",
    "source": "'lang'",
    "value": "lang",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`greek identifiers 3`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "́",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "String",
    "source": "'é'",
    "value": "é",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`greek identifiers 4`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "❌",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "1",
    "value": 1,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`identifiers and colons 1`] = `
[
  {
    "kind": "Identifier",
    "source": "foo",
  },
  {
    "kind": "Colon",
    "source": ":",
  },
  {
    "kind": "Identifier",
    "source": "bar",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`let hello = 1; 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "hello",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "1",
    "value": 1,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`mixed operators 1`] = `
[
  {
    "kind": "Identifier",
    "source": "a",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Identifier",
    "source": "b",
  },
  {
    "kind": "Colon",
    "source": ":",
  },
  {
    "kind": "Identifier",
    "source": "c",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`multiple identifiers 1`] = `
[
  {
    "kind": "Identifier",
    "source": "a",
  },
  {
    "kind": "Identifier",
    "source": "b",
  },
  {
    "kind": "Identifier",
    "source": "c",
  },
]
`;

exports[`string double quotes 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "s",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "String",
    "source": ""world"",
    "value": "world",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`string escapes 1`] = `
[
  {
    "kind": "String",
    "source": "'a\\nb'",
    "value": "a
b",
  },
]
`;

exports[`string single quotes 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "s",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "String",
    "source": "'hello'",
    "value": "hello",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`unicode identifiers 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "α",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "1",
    "value": 1,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
  {
    "kind": "Const",
    "source": "const",
  },
  {
    "kind": "Identifier",
    "source": "π",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "3.14",
    "value": 3.14,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`unicode identifiers 2`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "язык",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "String",
    "source": "'lang'",
    "value": "lang",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`unicode identifiers 3`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "́",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "String",
    "source": "'é'",
    "value": "é",
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`unicode identifiers 4`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "❌",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "1",
    "value": 1,
  },
  {
    "kind": "Semicolon",
    "source": ";",
  },
]
`;

exports[`unquoted identifier with digits 1`] = `
[
  {
    "kind": "Identifier",
    "source": "foo123",
  },
]
`;

exports[`whitespace and newlines 1`] = `
[
  {
    "kind": "Let",
    "source": "let",
  },
  {
    "kind": "Identifier",
    "source": "x",
  },
  {
    "kind": "Equals",
    "source": "=",
  },
  {
    "kind": "Number",
    "source": "1",
    "value": 1,
  },
]
`;
